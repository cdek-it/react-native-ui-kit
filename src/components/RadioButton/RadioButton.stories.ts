import type { Meta, StoryObj } from '@storybook/react'

import { RadioButton } from './RadioButton'

const meta: Meta<typeof RadioButton> = {
  title: 'RadioButton',
  component: RadioButton,
  args: { checked: false, disabled: false, state: 'default' },
  argTypes: { state: { control: 'radio', options: ['default', 'danger'] } },
}

export default meta

type Story = StoryObj<typeof RadioButton>

const RadioButtonStory: Story = { args: {} }
export { RadioButtonStory as RadioButton }
