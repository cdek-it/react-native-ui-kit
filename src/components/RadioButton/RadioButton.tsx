import { memo, useCallback, useMemo } from 'react'
import {
  type AccessibilityProps,
  Pressable,
  type PressableStateCallbackType,
  StyleSheet,
  View,
  type ViewProps,
} from 'react-native'
import Animated, { LinearTransition } from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

export interface RadioButtonProps
  extends AccessibilityProps,
    Pick<ViewProps, 'testID'> {
  /** Обработчик нажатия на кнопку */
  onPress: () => void
  /**
   * true, если необходим компонент в активном состоянии
   * @default false
   */
  checked?: boolean
  /**
   * Управление доступностью компонента
   * @default false
   */
  disabled?: boolean
  /** Выбор состояния компонента */
  state?: 'default' | 'danger'
}

export const RadioButton = memo<RadioButtonProps>(
  ({
    onPress,
    checked = false,
    disabled = false,
    state = 'default',
    testID,
    ...rest
  }) => {
    const styles = useStyles()

    const centerViewBackground = useMemo(
      () => [styles.defaultView, disabled && !checked && styles.disabledView],
      [disabled, checked, styles.defaultView, styles.disabledView]
    )

    const pressableStyles = useCallback(
      ({ pressed }: PressableStateCallbackType) => {
        const result = [styles.container, styles.default]

        if (checked) {
          result.push(styles.checked)
        }

        if (pressed) {
          result.push(styles.pressed)

          if (checked) {
            result.push(styles.checkedPressed)
          }
        }

        if (state === 'danger') {
          result.push(styles.danger)

          if (checked) {
            result.push(styles.dangerChecked)

            if (pressed) {
              result.push(styles.dangerCheckedPressed)
            }
          }
        }

        if (disabled) {
          result.push(styles.disabled)

          if (checked) {
            result.push(styles.disabledChecked)
          }
        }

        return StyleSheet.flatten(result)
      },
      [
        checked,
        disabled,
        state,
        styles.container,
        styles.default,
        styles.pressed,
        styles.checked,
        styles.checkedPressed,
        styles.danger,
        styles.dangerChecked,
        styles.dangerCheckedPressed,
        styles.disabled,
        styles.disabledChecked,
      ]
    )

    return (
      <>
        {!disabled && state === 'danger' && (
          <Animated.View
            layout={LinearTransition.duration(100)}
            style={styles.outline}
          />
        )}
        <Pressable
          disabled={disabled}
          style={pressableStyles}
          testID={testID || 'RadioButton_Pressable'}
          onPress={onPress}
          {...rest}
        >
          <View style={[styles.center, centerViewBackground]} />
        </Pressable>
      </>
    )
  }
)

const useStyles = makeStyles(({ theme }) => ({
  container: {
    position: 'absolute',
    width: theme.Form.RadioButton.radiobuttonWidth,
    height: theme.Form.RadioButton.radiobuttonHeight,
    borderRadius: theme.Form.RadioButton.radiobuttonWidth,
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: theme.General.focusShadowWidth,
    marginTop: theme.General.focusShadowWidth,
  },
  center: {
    width: theme.Form.RadioButton.radiobuttonIconSize,
    height: theme.Form.RadioButton.radiobuttonIconSize,
    borderRadius: theme.Form.RadioButton.radiobuttonIconSize,
  },
  outline: {
    position: 'absolute',
    width:
      theme.Form.RadioButton.radiobuttonWidth +
      theme.General.focusShadowWidth * 2,
    height:
      theme.Form.RadioButton.radiobuttonHeight +
      theme.General.focusShadowWidth * 2,
    borderRadius:
      theme.Form.RadioButton.radiobuttonHeight +
      theme.General.focusShadowWidth * 2,
    backgroundColor: theme.General.focusOutlineErrorColor,
  },

  // centerView
  defaultView: { backgroundColor: theme.Form.InputText.inputBg },
  disabledView: { backgroundColor: 'transparent' },

  // container viewState
  default: {
    borderColor: theme.Form.InputText.inputBorderColor,
    borderWidth: 1,
    backgroundColor: theme.Form.InputText.inputBg,
  },
  pressed: {
    borderColor: theme.Form.InputText.inputHoverBorderColor,
    backgroundColor: theme.Form.InputText.inputBg,
    borderWidth: 1,
  },
  checked: {
    borderColor: theme.Form.RadioButton.radiobuttonActiveBorderColor,
    backgroundColor: theme.Form.RadioButton.radiobuttonActiveBorderColor,
    borderWidth: 5,
  },
  checkedPressed: {
    borderColor: theme.Form.RadioButton.radiobuttonActiveHoverBorderColor,
    backgroundColor: theme.Form.RadioButton.radiobuttonActiveHoverBorderColor,
    borderWidth: 5,
  },
  // danger viewState
  danger: {
    borderColor: theme.Form.InputText.inputErrorBorderColor,
    backgroundColor: theme.Form.InputText.inputBg,
    borderWidth: 1,
  },
  dangerChecked: {
    borderColor: theme.Form.InputText.inputErrorBorderColor,
    backgroundColor: theme.Form.RadioButton.radiobuttonActiveBorderColor,
    borderWidth: 1,
  },
  dangerCheckedPressed: {
    borderColor: theme.Form.InputText.inputErrorBorderColor,
    backgroundColor: theme.Form.RadioButton.radiobuttonActiveBorderColor,
    borderWidth: 1,
  },
  // disabled viewState
  disabled: {
    borderColor: theme.Form.InputText.inputBorderColor,
    backgroundColor: theme.Button.Disabled.disabledButtonBg,
    opacity: 0.6,
    borderWidth: 1,
    mixBlendMode: 'luminosity',
  },
  disabledChecked: {
    borderColor: theme.Form.RadioButton.radiobuttonActiveBorderColor,
    backgroundColor: theme.Form.RadioButton.radiobuttonActiveBorderColor,
    borderWidth: 5,
  },
}))
