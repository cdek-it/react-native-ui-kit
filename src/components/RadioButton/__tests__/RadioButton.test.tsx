import { render } from '@testing-library/react-native'

import { RadioButton, type RadioButtonProps } from '../RadioButton'

describe('RadioButton', () => {
  const defaultProps: RadioButtonProps = {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onPress: () => {},
  }

  describe('snapshots', () => {
    const snapshotCases: Array<[string, Partial<RadioButtonProps>]> = [
      ['default', {}],
      [
        'checked = true, disabled = false, state = danger',
        { checked: true, disabled: false, state: 'danger' },
      ],
      [
        'checked = false, disabled = false, state = danger',
        { checked: false, disabled: false, state: 'danger' },
      ],

      [
        'checked = true, disabled = false, state = default',
        { checked: true, disabled: false, state: 'default' },
      ],
      [
        'checked = false, disabled = false, state = default',
        { checked: false, disabled: false, state: 'default' },
      ],

      [
        'checked = true, disabled = true, state = default',
        { checked: true, disabled: true, state: 'default' },
      ],
      [
        'checked = false, disabled = true, state = default',
        { checked: false, disabled: true, state: 'default' },
      ],
    ]

    test.each(snapshotCases)('%s', (_, props) => {
      const { toJSON } = render(<RadioButton {...defaultProps} {...props} />)

      expect(toJSON()).toMatchSnapshot()
    })
  })
})
