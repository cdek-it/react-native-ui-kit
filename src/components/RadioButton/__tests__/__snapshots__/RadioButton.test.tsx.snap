// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton snapshots checked = false, disabled = false, state = danger 1`] = `
[
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    layout={
      LinearTransition {
        "build": [Function],
        "durationV": 100,
        "randomizeDelay": false,
        "reduceMotionV": "system",
      }
    }
    style={
      {
        "backgroundColor": "#f6d4d1",
        "borderRadius": 28,
        "height": 28,
        "position": "absolute",
        "width": 28,
      }
    }
  />,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderColor": "#e26e63",
        "borderRadius": 21,
        "borderWidth": 1,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      }
    }
    testID="RadioButton_Pressable"
  >
    <View
      style={
        [
          {
            "borderRadius": 10.5,
            "height": 10.5,
            "width": 10.5,
          },
          [
            {
              "backgroundColor": "#ffffff",
            },
            false,
          ],
        ]
      }
    />
  </View>,
]
`;

exports[`RadioButton snapshots checked = false, disabled = false, state = default 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "borderColor": "#dbdada",
      "borderRadius": 21,
      "borderWidth": 1,
      "height": 21,
      "justifyContent": "center",
      "marginLeft": 3.5,
      "marginTop": 3.5,
      "position": "absolute",
      "width": 21,
    }
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</View>
`;

exports[`RadioButton snapshots checked = false, disabled = true, state = default 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.1000)",
      "borderColor": "#dbdada",
      "borderRadius": 21,
      "borderWidth": 1,
      "height": 21,
      "justifyContent": "center",
      "marginLeft": 3.5,
      "marginTop": 3.5,
      "mixBlendMode": "luminosity",
      "opacity": 0.6,
      "position": "absolute",
      "width": 21,
    }
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          {
            "backgroundColor": "transparent",
          },
        ],
      ]
    }
  />
</View>
`;

exports[`RadioButton snapshots checked = true, disabled = false, state = danger 1`] = `
[
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    layout={
      LinearTransition {
        "build": [Function],
        "durationV": 100,
        "randomizeDelay": false,
        "reduceMotionV": "system",
      }
    }
    style={
      {
        "backgroundColor": "#f6d4d1",
        "borderRadius": 28,
        "height": 28,
        "position": "absolute",
        "width": 28,
      }
    }
  />,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "borderColor": "#e26e63",
        "borderRadius": 21,
        "borderWidth": 1,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      }
    }
    testID="RadioButton_Pressable"
  >
    <View
      style={
        [
          {
            "borderRadius": 10.5,
            "height": 10.5,
            "width": 10.5,
          },
          [
            {
              "backgroundColor": "#ffffff",
            },
            false,
          ],
        ]
      }
    />
  </View>,
]
`;

exports[`RadioButton snapshots checked = true, disabled = false, state = default 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#188700",
      "borderColor": "#188700",
      "borderRadius": 21,
      "borderWidth": 5,
      "height": 21,
      "justifyContent": "center",
      "marginLeft": 3.5,
      "marginTop": 3.5,
      "position": "absolute",
      "width": 21,
    }
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</View>
`;

exports[`RadioButton snapshots checked = true, disabled = true, state = default 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#188700",
      "borderColor": "#188700",
      "borderRadius": 21,
      "borderWidth": 5,
      "height": 21,
      "justifyContent": "center",
      "marginLeft": 3.5,
      "marginTop": 3.5,
      "mixBlendMode": "luminosity",
      "opacity": 0.6,
      "position": "absolute",
      "width": 21,
    }
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</View>
`;

exports[`RadioButton snapshots default 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "borderColor": "#dbdada",
      "borderRadius": 21,
      "borderWidth": 1,
      "height": 21,
      "justifyContent": "center",
      "marginLeft": 3.5,
      "marginTop": 3.5,
      "position": "absolute",
      "width": 21,
    }
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</View>
`;
